File: Task solutions/Life/python/framework.py
from functools import reduce

def display(x):
  if isinstance(x, str):
    return str(x)
  else:
    try:
         return str(list(iter(x)))
    except TypeError:
         return str(x)

fail = "Test failed calling "

def arg_string(arguments) :
    args = display(arguments)
    return args[1:len(args)-1]

def test_function(function_name, expected, actual, args):
    if display(actual) != display(expected):
        print(fail + f"{function_name}({arg_string(args)}) Expected: {display(expected)}  Actual: {display(actual)}");
        raise TestFailure

def assert_true(function_name, args, actual, message) :
    if actual is not True :
        print(fail + f"{function_name}({arg_string(args)}) {message}")
        raise TestFailure

all_passed = "All tests passed"

def all_tests_passed():
    print(all_passed)

class TestFailure(Exception) : pass

File: Task solutions/Framework/csharp/Framework.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace Framework
{
    public static class Framework
    {
        public static string Display(object obj)
        {
            if (obj == null)  return null;
            if (obj is string) return $"{obj}";
            if (obj is Boolean) return (Boolean) obj ? "true" : "false";
            if (obj is IEnumerable)
            {
                var display = ((IEnumerable)obj).Cast<object>().Select(o => Display(o));
                return $@"{{{string.Join(',', display)}}}";
            }
            return obj.ToString();
        }

    public static string fail = "Test failed calling ";

    public static string ArgString(params object[] arguments) => arguments.Aggregate("", (s, a) => s + Display(a) + ", ").TrimEnd(' ', ',');

    public static void TestFunction(string functionName, object expected, object actual, params object[] args)
    {
        if (Display(actual) != Display(expected))
        {
            Console.WriteLine(fail + $"{functionName}({ArgString(args)}) Expected: {Display(expected)}  Actual: {Display(actual)}");
            throw new TestFailure();
        }
    }

    public static void AssertTrue(string functionName, string args, bool actual, string message)
    {
        if (actual != true)
        {
            Console.WriteLine(fail +$"{functionName}({ArgString(args)}) {message}");
            throw new TestFailure();
        }
    }

    public static string allTestsPassed = "All tests passed.";

    public static void AllTestsPassed()
    {
        Console.Write(allTestsPassed);
    }

    public class TestFailure : Exception { }

    }
}